lib/active_model/serializer/utils.rb

```diff
@@ -0,0 +1,42 @@
+module ActiveModel::Serializer::Utils
+  module_function
+
+    # converts the include hash to a standard format
+    # for constrained serialization recursion
+    #
+    # converts
+    #  [:author, :comments => [:author]] to
+    #  {:author => [], :comments => [:author]}
+    #
+    # and
+    #  [:author, :comments => {:author => :bio}, :posts => [:comments]] to
+    #  {:author => [], :comments => {:author => :bio}, :posts => [:comments]}
+    #
+    # The data passed in to this method should be an array where the last
+    # parameter is a hash
+    #
+    # the point of this method is to normalize the include
+    # options for the child relationships.
+    # if a sub inclusion is still an array after this method,
+    # it will get converted during the next iteration
+    def include_array_to_hash(include_array)
+      # still don't trust input
+      # but this also allows
+      #  include: :author syntax
+      include_array = Array[*include_array].compact
+
+      result = {}
+
+      hashes = include_array.select{|a| a.is_a?(Hash)}
+      non_hashes = include_array - hashes
+
+      hashes += non_hashes.map{ |association_name| { association_name => [] } }
```

Comment:
Maybe I'm just having a hard time following the code.  

the left side `hashes` is an array of hashes (`Hash` elements in `include_array`) and the right hand side is an array of hashes ( non-`Hash` elements in `include_array` mapped to an empty empty `Array`). `hashes = [{foo: :bar}]  + [{baz: []}] #=> [{:foo=>:bar}, {:baz=>[]}]` so I was wrong about `<<`, it should be `.concat` i.e. `hashes.concat(non_hashes.map.....)`

So, the transformation steps in the method are:

```ruby
include_array = [:author, comments: {author: :bio}, posts: [:comments]]

include_array =  Array[*include_array].compact
# => [:author, {:comments=>{:author=>:bio}, :posts=>[:comments]}]

result = {}

hashes = include_array.select{|a| a.is_a?(Hash)}
# => [{:comments=>{:author=>:bio}, :posts=>[:comments]}]

non_hashes = include_array - hashes
# => [:author]

hashes += non_hashes.map{ |association_name| { association_name => [] } }
# => [{:comments=>{:author=>:bio}, :posts=>[:comments]}, {:author=>[]}]

hashes.each{|hash| result.merge!(hash) }
# => [{:comments=>{:author=>:bio}, :posts=>[:comments]}, {:author=>[]}]

result
# => {:comments=>{:author=>:bio}, :posts=>[:comments], :author=>[]}
```

So, this is really

```ruby
include_array = [:author, comments: {author: :bio}, posts: [:comments]]
include_array.each_with_object({}) { |element, result|
  if element.is_a?(Hash)
    result.update(element)
  else
    result[element] = []
  end
}
```
right? so `include_array_to_hash` is really `hash_from_array(array,  non_hash_value: [])`

```ruby
def hash_from_array(array, non_hash_value: [])
  array.each_with_object({}) { |element, hash|
    if element.respond_to?(:keys)
      hash.update(element)
    else
      hash[element] = non_hash_value.dup
    end
  }
end
```

right?
  }
end